### PR TITLE
Assign name to Work Plugin when you register it

### PR DESCRIPTION
**Pull request by jsignell**
Closes: #3426

This is the most straightforward fix I could think of to make sure that the name sticks to the plugin. I'm not sure how bad it is to alter the plugin when you call register.

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -3961,6 +3961,8 @@ def register_worker_callbacks(self, setup=None):
         return self.register_worker_plugin(_WorkerSetupPlugin(setup))
 
     async def _register_worker_plugin(self, plugin=None, name=None):
+        if getattr(plugin, "name", None) is None:
+            plugin.name = name
```
Comment (mrocklin) [inline distributed/client.py]: Hrm, this seems a little janky.  We could also investigate changing `Scheduler.worker_plugins` from a list to a dict, or storing the plugins themselves as dictionaries like `{"plugin": my_plugin, "name": name}`.  Those probably both have problems (we lose ordering in the first case, and it's not clear what to do when we don't have a name).

Comment (mrocklin) [reply distributed/client.py]: Hrm, maybe not though.  I guess if it gets the job done?

Comment (quasiben) [reply distributed/client.py]: With the conditional check we don't need the additional `is None`

Comment (jsignell) [reply distributed/client.py]: @mrocklin - definitely janky :) but I didn't want to change existing behavior if the plugin has a name, so I think if we want to pass a dict, then we should use `plugin.name` as the default when name isn't provided as a kwarg. Which feels pretty similar to me. 

@quasiben - You don't think the name could be something that evaluates to False (like 0)?

Comment (quasiben) [reply distributed/client.py]: Oh! I wasn't thinking about numerical names.  Thanks for the correction

Comment (jsignell) [reply distributed/client.py]: So is this an ok solution? Should I go ahead and add/fix tests?

Comment (TomAugspurger) [reply distributed/client.py]: This seems OK to me. Changing the container from a list to dict also seems sensible, but probably will be better as a followup.

**Comment by TomAugspurger** [comment]
Sorry, one last question here: are we OK just mutating the user-provided object inplace? That seems a bit unfriendly.

Would it be better to require that WorkerPlugin have a name attribute set before they can be registered?

**Comment by jsignell** [comment]
Maybe this approach is unfriendly enough to be a bad idea and we should pursue the dict idea instead. Or maybe there shouldn't be a name kwarg on the register method at all. It is fundamentally kind of stange to allow two different methods for setting the name.

**Comment by jsignell** [comment]
@TomAugspurger so your proposal is to not have a name kwarg on the register and instead check that the plugin has a name and throw an error if it doesn't?

**Comment by TomAugspurger** [comment]
> @TomAugspurger so your proposal is to not have a name kwarg on the register and instead check that the plugin has a name and throw an error if it doesn't?

Yeah, though I haven't really thought through the implications of it, I was just thinking out loud.

If we think that requiring a `.name` is a good idea, I think we can deprecate the current behavior by allowing the name= keyword still, but warning if it's provided an the plugin doesn't have a name.

**Comment by nsmith-** [comment]
I think requiring a `.name` is better, less confusion having two ways to set it.
Would it make sense to have the base class declare it as a property such that if a user doesn't provide the attribute, there is some fallback? Then a user can even override the property if the want to use multiple instances of the same class as plugins.

**Comment by jsignell** [comment]
> Would it make sense to have the base class declare it as a property such that if a user doesn't provide the attribute, there is some fallback?

Is there a meaningful default name that we could set on the base class or would it just be an attr set to `None` such that the attr could be set on instances or overwritten by inherited classes?

**Comment by nsmith-** [comment]
I cannot think of anything more meaningful than `repr()` in the case that the user has not supplied a name either by attribute or by the argument to register, so I guess setting it to None seems the most simple.

**Comment by jsignell** [comment]
Ok sorry, this slipped off my radar for a bit. I've addressed the comments now.

**Comment by mrocklin** [comment]
> If we think that requiring a .name is a good idea, I think we can deprecate the current behavior by allowing the name= keyword still, but warning if it's provided an the plugin doesn't have a name.

Sorry to jump in here late.  As a reminder, a common use case is to have a user provide a callable function, rather than a class, and have Dask turn that into a plugin on the fly.  In my mind one of the major use cases of naming plugins is to replace old ones.

I think that before making a decision to remove the name parameter we probably need to gather a bit of information about how this gets used and why it was added in the first place.  I honestly don't recall this information, so some sleuthing and reporting back might be good here.

**Comment by jsignell** [comment]
Ok thanks for the context. I'll do some sleuthing.

**Comment by jsignell** [comment]
I am now actually using worker plugins and hopefully I understand them a bit better. 

I looked back at the original PR where worker plugins were added (#2453) and I don't understand why the naming behavior is how it is. Right now if you do:

```python
client.register_worker_plugin(PipInstall(packages=["pandas"])) 
```

and then do 

```python
client.register_worker_plugin(PipInstall(packages=["intake"])) 
```

you'll never get the second set of packages because the plugins have the same name. This is intentional according to the [docs](https://github.com/dask/distributed/blob/4205280f3f6584c59b5bc35af5f1938e328654f9/distributed/client.py#L4189:L4191). 

I'm closing this PR, since I can't reproduce the original issue on master anymore.

**Pull request closed without merging**